### PR TITLE
Added a new status : BACKLOG

### LINKED ISSUE #1855: Introducing BACKLOG status
### Issue Description
We've encountered situations where specific tasks are currently on hold or remain incomplete due to the  departure of their assigned individuals. Additionally, there are issues that, while valid, do not currently align with the immediate business priorities and can be deferred for future consideration. In response to these scenarios, there is a need to introduce a new status that precisely captures and communicates the nature of these tasks and issues.

### Current Behavior
There are only two options to keep for the above situations which do not explain the above task clearly .

### Expected Behavior 
After the code changes the Super user should be able to change these task's status to Backlog . Backlog will be the new status that will be applied to such tasks that are defined above . The scope of this issue is only limited to backend.

### PR DESCRIPTION
**Pull request by rishirishhh**
## Issue Ticket Number  1855
[https://github.com/Real-Dev-Squad/website-backend/issues/1855]

## Description

The purpose of this PR is to add a new status called BACKLOG.This status can only be applied by the superuser . There were certain changes done in the codebase as per this PR [https://github.com/Real-Dev-Squad/website-backend/pull/1756] that were removed because they were no longer needed . 

### Documentation Updated?

- [ ] Yes
- [X] No

### Under Feature Flag

- [X] Yes
- [ ] No

### Database Changes

- [X] Yes
- [ ] No

A new status is added called BACKLOG
### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [X] Yes
- [ ] No

## Screenshots

Screenshot from 2024-02-04 03-49-04

Screenshot from 2024-02-04 03-49-30

## Test Coverage
Screenshot from 2024-02-04 03-36-22

This closes #1855

### CODE REVIEW

File: controllers/tasks.js [hunk 1]
```diff
@@ -339,7 +327,11 @@ const updateTaskStatus = async (req, res, next) => {
 
     if (task.taskNotFound) return res.boom.notFound("Task doesn't exist");
     if (task.notAssignedToYou) return res.boom.forbidden("This task is not assigned to you");
-    if (task.taskData.status === TASK_STATUS.VERIFIED || req.body.status === TASK_STATUS.MERGED)
+    if (
+      task.taskData.status === TASK_STATUS.VERIFIED ||
+      req.body.status === TASK_STATUS.MERGED ||
+      req.body.status === TASK_STATUS.BACKLOG
```
Comment (sumitd94) [inline controllers/tasks.js:333]: NIT: we can destructure this `status` and use it here then

Comment (rishirishhh) [reply controllers/tasks.js:333]: Sure this will be done

**Comment by ajoykumardas12** [comment]
Some checkbox are not filled. Also, you can mention the issue number as `closes #num` which will automatically close the issue when the PR is merged.

**Review by sumitd94 — APPROVED** [review]

File: test/integration/tasks.test.js [hunk 2]
```diff
@@ -1460,87 +1470,4 @@ describe("Tasks", function () {
...
-
-      const updatedData = docSnapshot.data();
-      expect(updatedData.assignee).to.equal(null);
-      expect(updatedData.status).to.equal(TASK_STATUS.AVAILABLE);
-      expect(updatedData.percentCompleted).to.equal(0);
-      expect(updatedData.startedOn).to.equal(null);
-      expect(updatedData.endsOn).to.equal(null);
-    });
-
-    it("Should throw bad request if the req body is invalid", async function () {
```
Comment (ajoykumardas12) [inline test/integration/tasks.test.js:1528]: Is this test not needed anymore?

Comment (rishirishhh) [reply test/integration/tasks.test.js:1528]: No this test is not needed anymore as this test was written to test the piece of code that is no longer needed .

File: controllers/tasks.js [hunk 3]
```diff
@@ -334,38 +322,40 @@ const updateTaskStatus = async (req, res, next) => {
     let userStatusUpdate;
     const taskId = req.params.id;
     const { userStatusFlag } = req.query;
+    const { status } = req.body;
```
Comment (satyam73) [inline controllers/tasks.js:325]: is status is required field or option in body?

File: controllers/tasks.js [hunk 4]
```diff
@@ -334,38 +322,40 @@ const updateTaskStatus = async (req, res, next) => {
...
+      task.taskData.status === TASK_STATUS.VERIFIED ||
+      status === TASK_STATUS.MERGED ||
+      status === TASK_STATUS.BACKLOG
+    )
       return res.boom.forbidden("Status cannot be updated. Please contact admin.");
 
     if (userStatusFlag) {
       if (task.taskData.status === TASK_STATUS.DONE && req.body.percentCompleted < 100) {
-        if (req.body.status === TASK_STATUS.DONE || !req.body.status) {
+        if (status === TASK_STATUS.DONE || !status) {
```
Comment (satyam73) [inline controllers/tasks.js:340]: !status should be checked first so that JS program short circuits from that point onwards and it has not run the other condition

File: controllers/tasks.js [hunk 5]
```diff
@@ -334,38 +322,40 @@ const updateTaskStatus = async (req, res, next) => {
...
+      if ((status === TASK_STATUS.DONE || status === TASK_STATUS.VERIFIED) && task.taskData.percentCompleted !== 100) {
         if (req.body.percentCompleted !== 100) {
           return res.boom.badRequest("Status cannot be updated. Task is not done yet");
         }
       }
     }
 
     if (task.taskData.status === TASK_STATUS.COMPLETED && req.body.percentCompleted < 100) {
-      if (req.body.status === TASK_STATUS.COMPLETED || !req.body.status) {
+      if (status === TASK_STATUS.COMPLETED || !status) {
```
Comment (satyam73) [inline controllers/tasks.js:353]: !status should be checked first so that JS program short circuits from that point onwards and it has not run the other condition

File: test/integration/tasks.test.js (hunk 2 above)
Comment (satyam73) [reply test/integration/tasks.test.js:1528]: Is that part of code which is not need is removed from the codebase? Since it's not in the PR in my opinion, if removed please point me

File: test/integration/tasks.test.js [hunk 6]
```diff
@@ -1087,6 +1087,16 @@ describe("Tasks", function () {
 
       expect(res.body.message).to.be.equal("Status cannot be updated. Please contact admin.");
     });
+    it("Should give 403 if new status is 'BACKLOG' ", async function () {
+      taskId = (await tasks.updateTask({ ...taskData, assignee: appOwner.username })).taskId;
```
Comment (satyam73) [inline test/integration/tasks.test.js:1091]: Why updateTask model is being used here if not data will be updated?

File: test/integration/tasks.test.js
Comment (satyam73) [inline test/integration/tasks.test.js:1]: If status is `BACKLOG` who can update the task and who can't let me know please

**Review by satyam73 — CHANGES_REQUESTED** [review]

File: controllers/tasks.js (hunk 3 above)
Comment (satyam73) [reply controllers/tasks.js:325]: Since it's optional we can do this
const status = req.body?.status;

File: test/integration/tasks.test.js (hunk 2 above)
Comment (rishirishhh) [reply test/integration/tasks.test.js:1528]: yes it has been removed .

File: controllers/tasks.js [hunk 7]
```diff
@@ -289,18 +289,6 @@ const updateTask = async (req, res) => {
...
-    if (
-      requestData?.status === TASK_STATUS.AVAILABLE &&
-      task.taskData.status !== TASK_STATUS.AVAILABLE &&
-      Object.keys(req.body).length === 1
-    ) {
-      requestData.assignee = null;
-      requestData.percentCompleted = 0;
-      requestData.startedOn = null;
-      requestData.endsOn = null;
-    }
```
Comment (rishirishhh) [inline controllers/tasks.js:302]: @satyam73 this is the part of code that has been removed.

File: test/integration/tasks.test.js (hunk 2 above)
Comment (rishirishhh) [reply test/integration/tasks.test.js:1528]: https://github.com/Real-Dev-Squad/website-backend/pull/1875/files#r1478565929
@satyam73

File: controllers/tasks.js (hunk 3 above)
Comment (rishirishhh) [reply controllers/tasks.js]: changed this to the above stated way.

File: controllers/tasks.js (hunk 4 above)
Comment (rishirishhh) [reply controllers/tasks.js:340]: I have updated the way of extracting status

File: controllers/tasks.js (hunk 5 above)
Comment (rishirishhh) [reply controllers/tasks.js:353]: done

File: test/integration/tasks.test.js
Comment (rishirishhh) [reply test/integration/tasks.test.js:1]: This can be updated by superuser and cron jobs only

File: test/integration/tasks.test.js (hunk 6 above)
Comment (rishirishhh) [reply test/integration/tasks.test.js:1091]: the updateTask model seems is responsible for updating task data in the Firestore database based on certain conditions and handling dependencies. In the specific test case , it looks like the intention is to update the task data and obtain the updated taskId. The returned taskId is then used in the subsequent request to assert that the status cannot be updated to "BACKLOG". This test likely checks if the updateTask model correctly handles the case where the status is not allowed to be updated to "BACKLOG" under certain conditions.

File: test/integration/tasks.test.js [hunk 8]
```diff
@@ -1460,87 +1470,4 @@ describe("Tasks", function () {
...
-
-      const docSnapshot = await firestore.collection("tasks").doc("4kAkRv9TBlOfR6WEUhoQ").get();
-
-      const updatedData = docSnapshot.data();
-      expect(updatedData.assignee).to.equal(null);
-      expect(updatedData.status).to.equal(TASK_STATUS.AVAILABLE);
-      expect(updatedData.percentCompleted).to.equal(0);
-      expect(updatedData.startedOn).to.equal(null);
-      expect(updatedData.endsOn).to.equal(null);
-    });
```
Comment (Achintya-Chatterjee) [inline test/integration/tasks.test.js:1526]: why did you remove existing test cases from the codebase? It is intended?

Comment (rishirishhh) [reply test/integration/tasks.test.js:1526]: yes this is intended . 
This is the piece of code that was removed https://github.com/Real-Dev-Squad/website-backend/pull/1875/files#r1478565929

File: test/integration/tasks.test.js (hunk 6 above)
Comment (satyam73) [reply test/integration/tasks.test.js:1091]: If you want id of the task which has BACKLOG status you can simply create a fixture for this and add in the db in before hook or in this test itself why updating the task to get the id of the task?

Comment (rishirishhh) [reply test/integration/tasks.test.js:1091]: I used this approach because updating the task provides a test that covers the end-to-end behavior of the system, including interactions that occur during the task update process. Updating the task to obtain the ID creates a more realistic scenario where a task is being modified, and the test reflects the actual behavior of the system when updating a task.

**Review by satyam73 — APPROVED** [review]
LGTM

**Review by ajoykumardas12 — APPROVED** [review]
